### ISSUE TITLE
Regression from latest release with crate resolution

### ISSUE THREAD

**Issue by cmichi** [paritytech/scale-info#151]
This is likely a regression from https://github.com/paritytech/scale-info/pull/145, my hunch is that `::` is likely no longer always prepended to `scale_info`.

We have a test in ink! that started to fail since the new version of `scale-info` was published.

To reproduce:

Create a `Cargo.toml` with this content:
```toml
[package]
name = "foobar"
version = "0.1.0"
authors = ["[your_name] <[your_email]>"]
edition = "2021"

[dependencies]
ink_primitives = { version = "3.0", default-features = false }
ink_metadata = { version = "3.0", default-features = false, features = ["derive"], optional = true }
ink_env = { version = "3.0", default-features = false }
ink_storage = { version = "3.0", default-features = false }
ink_lang = { version = "3.0", default-features = false }

scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
scale-info = { version = "2", default-features = false, features = ["derive"], optional = true }

[lib]
name = "foobar"
path = "lib.rs"
crate-type = [
	# Used for normal contract Wasm blobs.
	"cdylib",
]

[features]
default = ["std"]
std = [
    "ink_metadata/std",
    "ink_env/std",
    "ink_storage/std",
    "ink_primitives/std",
    "scale/std",
    "scale-info/std",
]
ink-as-dependency = []
```

Crate a `lib.rs` with this content:
```rust
#![no_implicit_prelude]

#[::ink_lang::contract]
mod contract {
    #[ink(storage)]
    pub struct Contract {}

    impl Contract {
        #[ink(constructor)]
        pub fn constructor() -> Self {
            Self {}
        }

        #[ink(message)]
        pub fn message(&self) {}
    }
}
```

Then execute `cargo +nightly build`:

```
$ cargo +nightly build                                                                                                                               
   Compiling foobar v0.1.0 (/tmp/foobar)                                       
error[E0433]: failed to resolve: use of undeclared crate or module `scale_info`                                                                                
 --> lib.rs:6:5                                                                
  |                                                                            
6 |     pub struct Contract {}         
  |     ^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `scale_info`  
  |                                    
  = note: this error originates in the derive macro `::scale_info::TypeInfo` (in Nightly builds, run with -Z macro-backtrace for more info)                    
                                                                               
error[E0433]: failed to resolve: use of undeclared crate or module `scale_info` 
 --> lib.rs:6:5                        
  |                                                                                                                                                            
6 |     pub struct Contract {}                                                                                                                                 
  |     ^^^^^^^^^^^^^^^^^^^^^^ not found in `scale_info`                                                                                                       
  |
  = note: consider importing this struct:
          scale_info::Type
  = note: this error originates in the derive macro `::scale_info::TypeInfo` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: use of undeclared crate or module `scale_info` 
 --> lib.rs:6:5
  |
6 |     pub struct Contract {}
  |     ^^^^^^^^^^^^^^^^^^^^^^ not found in `scale_info`
  |
  = note: consider importing one of these items:
          scale_info::Path
          std::path::Path
  = note: this error originates in the derive macro `::scale_info::TypeInfo` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: use of undeclared crate or module `scale_info` 
 --> lib.rs:6:5
  |
6 |     pub struct Contract {}
  |     ^^^^^^^^^^^^^^^^^^^^^^ not found in `scale_info::build`
  |
  = note: consider importing this enum:
          scale_info::build::Fields
  = note: this error originates in the derive macro `::scale_info::TypeInfo` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `foobar` due to 8 previous errors
```

**Comment by jsdw** [comment]
I'd be interested to know what the expanded code is; ie how do the `ink` macros call the scale_info macro?

in the above example, `scale-info` is an optional dependency which hasn't been enabled, so I would expect that the `scale_info` crate wouldn't be found if the macro exapnds to include something like `#[derive(scale_info::TypeInfo)]` or similar.

**Comment by ascjones** [comment]
I've just copied a similar potential fix from `parity-scale-codec`. https://github.com/paritytech/parity-scale-codec/pull/328. I'll test it out on `ink! ` then see whether I can add a UI test for it.

**Comment by cmichi** [comment]
@jsdw See [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=69ee67d844ab94a160513bc70ad9c04b) for the `cargo expand` of our kinda-hello-world `flipper` example.

The crate is marked optional because the typical use case is to build the contracts for the `wasm32-unknown-unknown` target (`no_std`), whereas `scale-info` is only needed for some metadata generation that we do in `std`.

**Comment by ascjones** [comment]
https://github.com/paritytech/scale-info/pull/152 ready for review.

I originally skipped over it, but the reason for the failure in the example shown is the `#![no_implicit_prelude]`, which means that you need to use the fully qualified crate name e.g. `scale_info` will not do, must be `::scale_info`.

**Comment by ascjones** [comment]
Closed by #152
